Fixes #31017 - Do not use updates,security repos for Debian installations

quba42 opened the pull request:
https://projects.theforeman.org/issues/31017

Adding the line prevents it from defaulting to:
d-i apt-setup/services-select multiselect updates,security

quba42 committed:
Fixes #31017 - Do not use updates,security repos for Debian installations

https://projects.theforeman.org/issues/31017

Adding the line prevents it from defaulting to:
d-i apt-setup/services-select multiselect updates,security

ekohl reviewed (changes requested):
In the issue you state:
> This is never desirable when using Katello to provide content, and makes offline installations impossible.

However, not everyone is using Katello and this will be a huge degradation for them.

You probably want to at least detect if the Katello plugin is present, but perhaps really if Katello content is actually used. Not sure how you would do that.

quba42 commented:
> You probably want to at least detect if the Katello plugin is present, but perhaps really if Katello content is actually used. Not sure how you would do that.

Would a user defined boolean parameter guard (that defaults to false if unset) be a possible solution?
Or are we trying to avoid the inflationary use of boolean parameters in templates?

Like you, I don't know if there is some existing best practice way to check if Katello will be used for content...

ezr-ondrej commented:
> Would a user defined boolean parameter guard (that defaults to false if unset) be a possible solution?
> Or are we trying to avoid the inflationary use of boolean parameters in templates?

We try to avoid too much of template parameters, but if there is a reason to have one I believe we should add it.
Probably defaulting to false when katello is arround.
We could wrap the parameter fetch with method in [HostTemplate](https://github.com/theforeman/foreman/blob/cd4b1820bd38e888500f2b2f31b65f22857ade5e/lib/foreman/renderer/scope/macros/host_template.rb), that could be overriten in Katello if needed.

ezr-ondrej commented on app/views/unattended/provisioning_templates/provision/preseed_default.erb:
```diff
@@ -114,6 +114,7 @@ d-i apt-setup/local<%= repos %>/repository string <%= medium[:url] %> <%= @host.
 d-i apt-setup/local<%= repos %>/repository string http://debian.saltstack.com/debian <%= @host.operatingsystem.release_name %>-saltstack main
 d-i apt-setup/local<%= repos %>/comment string SaltStack Repository
 d-i apt-setup/local<%= repos %>/key string http://debian.saltstack.com/debian-salt-team-joehealy.gpg.key
+d-i apt-setup/services-select multiselect
```
Would this be better?
```suggestion
<% if plugin_present?('katello') %>
  d-i apt-setup/services-select multiselect
<% end %>
```

ekohl replied:
I'm not sure. Just because you installed Katello, does that mean that mirrored content is used? I don't think it is.

quba42 replied:
So I am making it depend on the parameter `kt_activation_keys` for now.
If that parameter does not meet the requirements, then the only thing I can think of is introducing a new parameter. Perhaps something like `katello_content_only`?

ezr-ondrej reviewed (changes requested):
Would it be ok to add the katello safeguard?

quba42 commented:
Thanks for the continued input. This PR is still on my TODO list, but other priorities keep getting in the way...

quba42 commented:
So I think I have found a preexisting parameter that captures the desired condition exactly:
`kt_activation_keys` is used to set the activation key hosts should use to receive content via Katello. Therefore it should only ever be set for hosts that receive their content from Katello?
Do you think this would work @ekohl?

Note: I also removed the line I added from the salt block (it was never my intention to place it there, a complete accident/oversight).

ekohl commented on app/views/unattended/provisioning_templates/provision/preseed_default.erb:
```diff
@@ -125,6 +125,13 @@ d-i apt-setup/local<%= repos %>/key string http://keyserver.ubuntu.com/pks/looku
+# security repos during the initial installation. Prevents us from defaulting to:
+# d-i apt-setup/services-select multiselect updates,security
+d-i apt-setup/services-select multiselect
+<% end -%>
+
```
If you write it like this, you avoid an empty line in the generated output.
```suggestion

<% end -%>
```

quba42 replied:
Applied.

ekohl reviewed (changes requested):
Small nit inline, otherwise :+1: for the logic. I think the nit may help you to avoid regenerating the snapshots (which will fail the tests).

ekohl reviewed (approved):
:+1: if Jenkins agrees.

ekohl commented:
ok to test

quba42 commented:
Jenkins is sending mixed messages: "178 tests run, 0 skipped, 0 failed. => RED!"

ekohl commented:
[test katello]

quba42 commented:
@ekohl, @ezr-ondrej I don't know how to continue with this. The Katello test result (184 tests run, 0 skipped, 0 failed. => RED!) does not make much sense to me, and does not give me anything to go on. Any ideas?

ezr-ondrej commented:
[test katello] I think it should be fixed, it was not your fault, the test suite was broken :(

ezr-ondrej reviewed (approved):
Thanks a lot @quba42, nice addition!

ezr-ondrej commented:
Thanks @ekohl for the review and valuable input :+1:
